contracts/src/v2/Calls.sol

```diff
@@ -0,0 +1,145 @@
+import {Call} from "../utils/Call.sol";
+import {Token} from "../Token.sol";
+import {Upgrade} from "../Upgrade.sol";
+import {Functions} from "../Functions.sol";
+import {Constants} from "../Constants.sol";
+
+import {Ticket, TransferKind, OperatingMode} from "./Types.sol";
+
+import {UD60x18, ud60x18, convert} from "prb/math/src/UD60x18.sol";
+
+/// @title Library for implementing Ethereum->Polkadot ERC20 transfers.
+library CallsV2 {
+    using Address for address;
+    using SafeTokenTransfer for IERC20;
+    using SafeNativeTransfer for address payable;
+
+    error InvalidProof();
+    error InvalidNonce();
+    error NotEnoughGas();
+    error FeePaymentToLow();
+    error Unauthorized();
+    error Disabled();
+    error AgentAlreadyCreated();
+    error AgentDoesNotExist();
+    error ChannelAlreadyCreated();
+    error ChannelDoesNotExist();
+    error InvalidChannelUpdate();
+    error AgentExecutionFailed(bytes returndata);
+    error InvalidAgentExecutionPayload();
+    error InvalidConstructorParams();
+    error AlreadyInitialized();
+    error TokenNotRegistered();
+
+    error InvalidAsset();
+
+    address public constant WETH_ADDRESS = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+
+    // Send an XCM with assets to Polkadot Asset Hub
+    //
+    // Params:
+    //   * `xcm` (bytes): SCALE-encoded XCM message
+    //   * `assets` (bytes[]): Array of asset transfer instructions
+    //
+    // The specified assets will be locked/burned locally, and their foreign equivalents will
+    // be minted/unlocked on Polkadot Asset Hub.
+    //
+    // Supported asset instructions:
+    // * ERC20: abi.encode(0, tokenAddress, value)
+    //
+    function sendMessage(bytes calldata xcm, bytes[] calldata assets) external {
+        bytes[] memory encodedAssets = new bytes[](assets.length);
+        for (uint256 i = 0; i < assets.length; i++) {
+            encodedAssets[i] = _handleAsset(assets[i]);
+        }
+
+        Ticket memory ticket =
+            Ticket({origin: msg.sender, assets: encodedAssets, xcm: xcm});
+        _submitOutbound(ticket);
+    }
+
+    /*
+    * Internal functions
+    */
+
+    // Submit an outbound message to Polkadot, after taking fees
+    function _submitOutbound(Ticket memory ticket) internal {
+        CoreStorage.Layout storage $ = CoreStorage.layout();
+
+        // Ensure outbound messaging is allowed
+        _ensureOutboundMessagingEnabled();
+
+        // Wrap provided ether and transfer to AssetHub agent
+        address assetHubAgent = Functions.ensureAgent(Constants.ASSET_HUB_AGENT_ID);
+        WETH9(payable(WETH_ADDRESS)).deposit{value: msg.value}();
+        IERC20(WETH_ADDRESS).safeTransfer(assetHubAgent, msg.value);
+
+        $.outboundNonce = $.outboundNonce + 1;
+
+        bytes memory payload =
+            SubstrateTypes.encodePayloadV2(ticket.origin, ticket.assets, ticket.xcm);
+
+        emit IGateway.OutboundMessageAccepted($.outboundNonce, msg.value, payload);
```

Comment:
Cool, A lot more simple now!

With `ticket.xcm` does that mean for V2 there is no need to construct xcm on chain any more?

It's the responsibility of relayer to validate the xcm and check profitable or not.

On BH we just decode the xcm with
```VersionedXcm::<()>::decode_with_depth_limit(
						MAX_XCM_DECODE_DEPTH,
						&mut data,
					)
``` and forward it to AH transparently.